### PR TITLE
Fix DNS-SD behavior relating to commissioning

### LINKED ISSUE #8711: Node to advertise non-commissionable mode in the DNS advertisement
At present device advertises commissionable mode even after device is provisioned i.e. sub type C is set to 1, instead of 0.

$ avahi-browse -t _services._dns-sd._udp | grep _matterc._udp

wlan0 IPv6 _C1 _sub._matterc._udp local
wlan0 IPv6 _L3840 _sub._matterc._udp local
wlan0 IPv6 _S0 _sub._matterc._udp local
wlan0 IPv6 _T257 _sub._matterc._udp local
wlan0 IPv6 _V9050 _sub._matterc._udp local
eth0 IPv6 _C1 _sub._matterc._udp local
eth0 IPv6 _L3840 _sub._matterc._udp local
eth0 IPv6 _S0 _sub._matterc._udp local
eth0 IPv6 _T257 _sub._matterc._udp local
eth0 IPv6 _V9050 _sub._matterc._udp local
eth0 IPv4 _C1 _sub._matterc._udp local
eth0 IPv4 _L3840 _sub._matterc._udp local
eth0 IPv4 _S0 _sub._matterc._udp local
eth0 IPv4 _T257 _sub._matterc._udp local
eth0 IPv4 _V9050 _sub._matterc._udp local
docker0 IPv4 _V9050 _sub._matterc._udp local
docker0 IPv4 _T257 _sub._matterc._udp local
docker0 IPv4 _S0 _sub._matterc._udp local
docker0 IPv4 _L3840 _sub._matterc._udp local
docker0 IPv4 _C1 _sub._matterc._udp local

per TC https://github.com/chip-csg/chip-test-plans/blob/master/src/securechannel.adoc#341-tc-sc-41-commissionable-node-discovery---commissionee-case, need an API to put the device in non-commissionable mode.

### PR DESCRIPTION
**Pull request by chrisdecenzo**
#### Problem
* Fixes #8711  - Node to advertise non-commissionable mode in the DNS advertisement

#### Change overview
Restart Mdns at various points in the device lifecycle to reflect changes in commissioning mode
Fix CM=0/1 and AP=0/1 key/value pairs in commissionable node discovery DNS-SD response
Added shell command to linux apps to allow manual changes to commissioning mode

Usage: commissionee restartmdns <commissioningMode> <additionalPairing>

#### Testing
* Run lighting-app with shell enabled, and chip-tool to print discovery information

1) start uncommissioned lighting-app with shell enabled
2) chip-tool discover commissionables (check CM=0 and AC=1 values)
3) commission lighting-app successfully using chip-tool
4) chip-tool discover commissionables (check CM=0 and AC=0 values)
5) from lighting-app shell, turn on commissioning mode: "commissionee restartmdns true true"
6) chip-tool discover commissionables (check CM=1 and AC=1 values)
7) from lighting-app shell, turn off commissioning mode: "commissionee restartmdns false false"
8) chip-tool discover commissionables (check CM=0 and AC=0 values)

### CODE REVIEW

**Review by pan-apple — APPROVED** [review]

**Review by cecille — APPROVED** [review]

**Review by sharadb-amazon — APPROVED** [review]

File: src/app/server/Mdns.cpp [hunk 1]
```diff
@@ -122,28 +122,23 @@ CHIP_ERROR AdvertiseOperational()
...
 {
-    return Advertise(true);
+    return Advertise(true, commissioningMode, additionalCommissioning);
 }
 
-/// commissionableNode
-// CHIP_ERROR Advertise(chip::Mdns::CommssionAdvertiseMode mode)
-CHIP_ERROR Advertise(bool commissionableNode)
+/// Set MDNS commissioner advertisement or commissionable node advertisement
+CHIP_ERROR Advertise(bool commissionableNode, bool commissioningMode, bool additionalCommissioning)
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:135]: The callsites of this are pretty unreadable... Should this take a bitfield arg instead, so callsites would list the bits that are true and it would be clearer what is being advertised?

File: examples/platform/linux/CommissioneeShellCommands.cpp [hunk 2]
```diff
@@ -84,6 +88,21 @@ static CHIP_ERROR CommissioneeHandler(int argc, char ** argv)
         return error  = SendUDC(true, chip::Transport::PeerAddress::UDP(commissioner, port));
     }
 #endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    else if (strcmp(argv[0], "restartmdns") == 0)
+    {
+        bool commissioningMode = false;
+        bool additionalPairing = false;
```
Comment (bzbarsky-apple) [inline examples/platform/linux/CommissioneeShellCommands.cpp:94]:
```suggestion
        bool additionalCommissioning = false;
```

File: src/app/server/Mdns.cpp (hunk 1 above)
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:135]: Also, should that `additionalCommissioning` arg be `notInitialCommissioning`, or perhaps reversed and `initialCommissioning`?  It's not 100% clear what "additional" means....

File: src/app/server/Mdns.cpp [hunk 3]
```diff
@@ -247,22 +242,27 @@ CHIP_ERROR Advertise(bool commissionableNode)
 /// (Re-)starts the minmdns server
 void StartServer()
 {
-    ChipLogProgress(Discovery, "Start dns-sd server");
+    StartServer(false, false);
+}
+
+void StartServer(bool commissioningMode, bool additionalCommissioning)
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:248]: Again, the callsites are pretty inscrutable.....

File: src/app/server/Mdns.cpp [hunk 4]
```diff
@@ -247,22 +242,27 @@ CHIP_ERROR Advertise(bool commissionableNode)
 /// (Re-)starts the minmdns server
 void StartServer()
 {
-    ChipLogProgress(Discovery, "Start dns-sd server");
+    StartServer(false, false);
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:245]: Is this the right behavior for `StartServer` callers?  Should we perhaps not have this signature at all?  Seems easy to misuse....

File: src/app/server/RendezvousServer.cpp [hunk 5]
```diff
@@ -94,6 +96,9 @@ CHIP_ERROR RendezvousServer::WaitForPairing(const RendezvousParameters & params,
         ReturnErrorOnFailure(GetAdvertisementDelegate()->StartAdvertisement());
     }
 
+    // reset all advertising
+    app::Mdns::StartServer(true, true);
```
Comment (bzbarsky-apple) [inline src/app/server/RendezvousServer.cpp:100]: Why is additionalCommissioning true here?

File: src/app/server/Mdns.cpp [hunk 6]
```diff
@@ -247,22 +242,27 @@ CHIP_ERROR Advertise(bool commissionableNode)
...
+void StartServer(bool commissioningMode, bool additionalCommissioning)
+{
     CHIP_ERROR err = chip::Mdns::ServiceAdvertiser::Instance().Start(&chip::DeviceLayer::InetLayer, chip::Mdns::kMdnsPort);
 
-    // TODO: advertise this only when really operational once we support both
-    // operational and commisioning advertising is supported.
+    // TODO: advertise all operational ids once we support multiple
     if (GetCurrentNodeId() != kUndefinedNodeId)
     {
+        ChipLogProgress(Discovery, "Start dns-sd server CM=%d, AC=%d", commissioningMode ? 1 : 0, additionalCommissioning ? 1 : 0);
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:255]: These are the _A1 and _C1 bits from the spec, right?  Can we just label them that way in the logs?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/platform/linux/CommissioneeShellCommands.cpp [hunk 7]
```diff
@@ -84,6 +88,21 @@ static CHIP_ERROR CommissioneeHandler(int argc, char ** argv)
...
+        if (strcmp(argv[1], "true") == 0)
```
Comment (andy31415) [inline examples/platform/linux/CommissioneeShellCommands.cpp:95]: is argc guaranteed to always be at least 3?

Wondering if this is easier to read:

```cpp
bool commisioningMode      =  (argc > 1) && (strcmp(argv[1], "true") == 0);
bool AdditionalComisioning =  (argc > 2) && (strcmp(argv[2], "true") == 0);
```

File: src/app/server/Mdns.cpp [hunk 8]
```diff
@@ -122,28 +122,23 @@ CHIP_ERROR AdvertiseOperational()
 /// Set MDNS commissioner advertisement
 CHIP_ERROR AdvertiseCommisioner()
 {
-    return Advertise(false);
+    return Advertise(false, false, false);
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:125]: Could we use a bitfield here? figuring out what true/false is seems really hard to follow. We have BitFlags for such things.

Comment (Damian-Nordic) [reply src/app/server/Mdns.cpp:125]: Or a simple enum like:
```
enum class CommissioningMode
{
  kDisabled,
  kEnabled,
  kEnabledWithAdminCommand
};

File: examples/platform/linux/CommissioneeShellCommands.cpp (hunk 7 above)
Comment (andy31415) [reply examples/platform/linux/CommissioneeShellCommands.cpp:95]: Just realized that this is not the same as additional stays false if commisioning mode is false.

I believe this makes a stronger case of bitfields and the argument here should be just one mode, because the pair "true/false" is invalid.

File: src/app/server/Mdns.cpp (hunk 8 above)
Comment (andy31415) [reply src/app/server/Mdns.cpp:125]: Was also thinking to consider the previous bool. Looking at `Advertise(false)` even as it stands today, what does it mean? Seems to somehow hint of 'do/do not advertise' but still somewhat vague.

File: src/app/server/RendezvousServer.cpp (hunk 5 above)
Comment (tcarmelveilleux) [reply src/app/server/RendezvousServer.cpp:100]: StartServer seems like an overall incorrect name. "StartResponder" is more like it. Furthermore, starting/stopping advertisement is separate from starting/stopping an  mDNS responder.

For instance: Thread nodes neither use mDNS (they use unicast DNS to SRP) nor run a server (they register to SRP on border router). This legacy naming from earlier is starting to make the code much harder to read/understand.

File: src/app/server/Mdns.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: added bitmap

File: examples/platform/linux/CommissioneeShellCommands.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply examples/platform/linux/CommissioneeShellCommands.cpp]: fixed

File: src/app/server/Mdns.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: changed to bitmap and added comments on flag definitions

File: src/app/server/Mdns.cpp (hunk 3 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: replaced with bitmaps

File: src/app/server/Mdns.cpp (hunk 4 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: removed StartServer()

File: src/app/server/RendezvousServer.cpp (hunk 5 above)
Comment (chrisdecenzo) [reply src/app/server/RendezvousServer.cpp]: added comment to code regarding why these arguments
I prefer not to change the name of the method in this PR though happy to take that in a future one.

File: src/app/server/Mdns.cpp (hunk 8 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: I think enum is better than bitfield

File: src/app/server/Mdns.cpp (hunk 6 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: these are the _A1 is the query value while AC=1 is the key/value pair, so both are in the spec.

File: src/app/server/Mdns.h [hunk 9]
```diff
@@ -19,26 +19,34 @@
...
-// CHIP_ERROR Advertise(chip::Mdns::CommssionAdvertiseMode mode);
-CHIP_ERROR Advertise(bool commissionableNode);
+CHIP_ERROR AdvertiseCommissionableNode(BitFlags<KeyValueFlags> flags);
 
 /// (Re-)starts the minmdns server
-void StartServer();
+//
+// NOTE: when device has never been commissioned, kAdditionalCommissioning flag will be ignored
+//
+void StartServer(BitFlags<KeyValueFlags> flags);
```
Comment (tcarmelveilleux) [inline src/app/server/Mdns.h:49]: Why not simply have StartServer() and immediately follow by Advertise? Why couple StartServer and AdvertiseCommissionableNode ?

File: src/app/server/Mdns.cpp (hunk 6 above)
Comment (bzbarsky-apple) [reply src/app/server/Mdns.cpp]: There is no AC=1 in the spec.  There's AP=1....

File: src/app/server/Mdns.cpp (hunk 8 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: ok, changed it to an enum since I needed to fix a build breakage

File: examples/platform/linux/CommissioneeShellCommands.cpp (hunk 7 above)
Comment (chrisdecenzo) [reply examples/platform/linux/CommissioneeShellCommands.cpp]: changed to enum from bitfield

File: src/app/server/Mdns.h (hunk 9 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: I think that makes sense but my intention here was just to fix a few bugs in the DNS-SD parameters and not re-visit the current design for method invocation.

File: examples/platform/linux/CommissioneeShellCommands.cpp [hunk 10]
```diff
@@ -84,6 +88,27 @@ static CHIP_ERROR CommissioneeHandler(int argc, char ** argv)
         return error  = SendUDC(true, chip::Transport::PeerAddress::UDP(commissioner, port));
     }
 #endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    else if (strcmp(argv[0], "restartmdns") == 0)
+    {
+        // if commissioning mode is true
+        if ((argc > 1) && (strcmp(argv[1], "true") == 0))
```
Comment (andy31415) [inline examples/platform/linux/CommissioneeShellCommands.cpp:94]: Since this is only  3 values, the pair "false true" is invalid.

How about making argv[1]  one of "disabled, enabled, enabled_with_additional" or similar naming?

File: examples/platform/linux/CommissioneeShellCommands.cpp [hunk 11]
```diff
@@ -61,6 +62,9 @@ static CHIP_ERROR PrintAllCommands()
     streamer_printf(sout,
                     "  sendudc <address> <port>   Send UDC message to address. Usage: commissionee sendudc 127.0.0.1 5543\r\n");
 #endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    streamer_printf(sout,
+                    "  restartmdns <commissioningMode> <additionalPairing>   Start Mdns with given settings. Usage: commissionee "
+                    "startmdns true true\r\n");
```
Comment (andy31415) [inline examples/platform/linux/CommissioneeShellCommands.cpp:67]: nit: example seems to be missing a "re".

File: src/app/server/Mdns.cpp [hunk 12]
```diff
@@ -119,31 +119,19 @@ CHIP_ERROR AdvertiseOperational()
...
     auto advertiseParameters = chip::Mdns::CommissionAdvertisingParameters().SetPort(CHIP_PORT).EnableIpV4(true);
     advertiseParameters.SetCommissionAdvertiseMode(commissionableNode ? chip::Mdns::CommssionAdvertiseMode::kCommissionableNode
                                                                       : chip::Mdns::CommssionAdvertiseMode::kCommissioner);
 
-    // TODO: device can re-enter commissioning mode after being fully provisioned
-    // (additionalPairing == true)
-    bool notYetCommissioned = !DeviceLayer::ConfigurationMgr().IsFullyProvisioned();
-    bool additionalPairing  = false;
-    advertiseParameters.SetCommissioningMode(notYetCommissioned, additionalPairing);
+    advertiseParameters.SetCommissioningMode(commissioningMode, additionalCommissioning);
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:134]: We fix the 'setCommisioningMode' to also not take several bools as arguments. Like split to 'EnableCommisioning' and 'EnableAdditionalCommisioning'. I am unclear why these assignment operations are made into a single call. `SetFoo(true, false)` is much less readable than `SetBar(true); SetBaz(false);`

I realize this is existing code, however given small change I would rather boy/girl-scout approach here and fix the advertise parameter awkwardness as  well. Multi-bool parameters make my code review  harder (I have to validate that the order of bool params matches the order that the header declares and uses them).

File: src/app/server/Mdns.cpp [hunk 13]
```diff
@@ -244,25 +232,50 @@ CHIP_ERROR Advertise(bool commissionableNode)
     return mdnsAdvertiser.Advertise(advertiseParameters);
 }
 
+/// Set MDNS commissioner advertisement
+CHIP_ERROR AdvertiseCommisioner()
+{
+    return Advertise(false, CommissioningMode::kDisabled);
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:238]: add a comment of  what false is here and true below (like `Advertise(false /* commisionableNode */, ...)`).

Bool arguments are often non-obvious. Code reading I would expect 'Advertise(false)' to mean "do not advertise". Can we fix this readability problem?

File: src/app/server/Mdns.h [hunk 14]
```diff
@@ -19,26 +19,37 @@
...
 
 /// Set MDNS commissionable node advertisement
-CHIP_ERROR AdvertiseCommissionableNode();
+CHIP_ERROR AdvertiseCommissionableNode(CommissioningMode mode);
 
-/// Set MDNS advertisement
-// CHIP_ERROR Advertise(chip::Mdns::CommssionAdvertiseMode mode);
-CHIP_ERROR Advertise(bool commissionableNode);
+/// (Re-)starts the minmdns server using default state
+// Commissioning mode is enabled if device has not yet been commissioned
```
Comment (andy31415) [inline src/app/server/Mdns.h:45]: Please use `///` to continue the doc comment for consistency. I am unsure if doxygen keeps following // and generally using 3 everywhere seems to make things align somewhat nicer.

File: src/app/server/Mdns.cpp [hunk 15]
```diff
@@ -119,31 +119,19 @@ CHIP_ERROR AdvertiseOperational()
...
-CHIP_ERROR AdvertiseCommisioner()
+/// Overloaded utility method for commissioner and commissionable advertisement
+// This method is used for both commissioner discovery and commissionable node discovery since
+// they share many fields.
+CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
 {
-    return Advertise(false);
-}
+    bool commissioningMode = (mode == CommissioningMode::kEnabled || mode == CommissioningMode::kEnabledAsAdditionalCommissioning);
+    bool additionalCommissioning = (mode == CommissioningMode::kEnabledAsAdditionalCommissioning);
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:128]: As a followup: we should make these APIs clearer so someone does not "fix" additionalCommissioning to be based on whether we already have an opcert.  That's not what it means.

File: src/app/server/Mdns.cpp [hunk 16]
```diff
@@ -244,25 +232,50 @@ CHIP_ERROR Advertise(bool commissionableNode)
...
+// NOTE: when device has never been commissioned, mode will be treated as CommissioningMode::kEnabled
+//
+void StartServer(CommissioningMode mode)
+{
     CHIP_ERROR err = chip::Mdns::ServiceAdvertiser::Instance().Start(&chip::DeviceLayer::InetLayer, chip::Mdns::kMdnsPort);
 
-    // TODO: advertise this only when really operational once we support both
-    // operational and commisioning advertising is supported.
+    // TODO: advertise all operational ids once we support multiple
     if (GetCurrentNodeId() != kUndefinedNodeId)
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:263]: This needs a merge to tip...

File: src/app/server/Mdns.h [hunk 17]
```diff
@@ -19,26 +19,37 @@
 
 #include <core/CHIPError.h>
 #include <stddef.h>
+#include <support/BitFlags.h>
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.h:22]: Bitflags is not used, right?

File: src/app/server/Mdns.h [hunk 18]
```diff
@@ -19,26 +19,37 @@
...
 
 namespace chip {
 namespace app {
 namespace Mdns {
 
+enum class CommissioningMode
+{
+    kDisabled,
+    kEnabled,
+    kEnabledAsAdditionalCommissioning
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.h:32]: It's worth documenting when one would use which of these; `kEnabledAsAdditionalCommissioning` means that we are in commissioning mode as a result of the "Open Commissioning Window" command or "Open Basic Commissioning Window" command and must not be set in any other cases, right?

File: src/app/server/Mdns.h [hunk 19]
```diff
@@ -19,26 +19,37 @@
...
-// CHIP_ERROR Advertise(chip::Mdns::CommssionAdvertiseMode mode);
-CHIP_ERROR Advertise(bool commissionableNode);
+/// (Re-)starts the minmdns server using default state
+// Commissioning mode is enabled if device has not yet been commissioned
+void StartServer();
 
 /// (Re-)starts the minmdns server
-void StartServer();
+//
+// NOTE: when device has never been commissioned, kAdditionalCommissioning flag will be ignored
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.h:50]: Confusing comment, because there is no such flag....  Please have this refer to the actual value that will be "ignored" and say what "ignored" means: will we not advertise commissioning mode at all (because someone is just passing random values in), or will we advertise it but without the AC=1/_A bits?   I'd somewhat prefer the former....

File: src/app/server/Mdns.cpp [hunk 20]
```diff
@@ -244,25 +232,50 @@ CHIP_ERROR Advertise(bool commissionableNode)
...
+
+/// Set MDNS commissionable node advertisement
+CHIP_ERROR AdvertiseCommissionableNode(CommissioningMode mode)
+{
+    return Advertise(true, mode);
+}
+
 /// (Re-)starts the minmdns server
+// Commissioning mode is enabled if device has not yet been commissioned
 void StartServer()
```
Comment (bzbarsky-apple) [inline src/app/server/Mdns.cpp:249]: If we're keeping this signature, why the changes to its various callers to call `StartServer(CommissioningMode::kDisabled)` directly?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — CHANGES_REQUESTED** [review]
Marking as 'change requested' to fix up the (bool,bool) arguments - I feel strongly that we allowed that  to go in and need to fix it. Too much bitrot that makes reviewing/maintainance harder.

Should be a small change.

File: examples/platform/linux/CommissioneeShellCommands.cpp (hunk 11 above)
Comment (chrisdecenzo) [reply examples/platform/linux/CommissioneeShellCommands.cpp]: fixed

File: src/app/server/Mdns.cpp (hunk 12 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: fixed

File: src/app/server/Mdns.h (hunk 14 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: fixed

File: src/app/server/Mdns.cpp (hunk 13 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: fixed

File: src/app/server/Mdns.h (hunk 17 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: fixed

File: src/app/server/Mdns.h (hunk 18 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: added comments

File: src/app/server/Mdns.h (hunk 19 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: cleaned up

File: src/app/server/Mdns.cpp (hunk 20 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp:254]: reverted Server.cpp usage to previous signature

File: src/app/server/Mdns.cpp (hunk 15 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp:128]: > Marking as 'change requested' to fix up the (bool,bool) arguments - I feel strongly that we allowed that to go in and need to fix it. Too much bitrot that makes reviewing/maintainance harder.
> 
> Should be a small change.

ok, please take a look at the latest thanks

File: src/app/server/Mdns.cpp (hunk 16 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: done

File: src/app/server/Mdns.cpp (hunk 15 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp:124]: I've added many comments to the use of these fields. I think it would be useful to have a global state somewhere which tracks commissioning mode for the device (and whether its basic or enhanced). That would allow us to remove these arguments completely.

Comment (msandstedt) [reply src/app/server/Mdns.cpp:124]: Isn't `RendezvousServer` the entity that should push this state to Mdns?  It by definition knows whether the node is commissionable or not.  I don't think the Mdns code is well-suited to pull this state from some central location instead.  Much of the 'what-state-are-we-in' logic in the configuration manager is broken anyway.

I think the problem is more that we've embedded all sorts of inference logic in the Mdns code.  Why is Advertise called from `StartServer` instead of `RendezvousServer`?  Why does the Mdns code need to infer what should be advertised on behalf of another module in the sdk?

It would be nice if the Mdns code provided public methods like these, but didn't actually call them: `AdvertiseCommissionableNode`, `AdvertiseCommissionableNodeAP`, `SuspendCommissionableNodeAdvertisement`.

If these were instead called from `RendezvousServer`, I think the Mdns code would be much simpler.

File: src/app/server/Mdns.h [hunk 21]
```diff
@@ -24,21 +24,31 @@ namespace chip {
...
+/// (Re-)starts the minmdns server using default state
+/// - if device has not yet been commissioned, then commissioning mode will show as enabled (CM=1, AC=0)
+/// - if devica has been commissioned, then commissioning mode will show as disabled (CM=0, AC=0)
+void StartServer();
 
 /// (Re-)starts the minmdns server
-void StartServer();
+/// - if device has not yet been commissioned, then commissioning mode will show as enabled (CM=1, AC=0)
+/// - if devica has been commissioned, then commissioning mode will reflect the state of mode argument
+void StartServer(CommissioningMode mode);
```
Comment (andy31415) [inline src/app/server/Mdns.h:51]: Why do we have two StartServer commands? if the default is CM=0, AC=0, why not setting mode to the default of the startserver without arguments? that  would make the header clearer. Just keeping:

```cpp
void StartServer(CommisioningMode mode = CommissioningMode::kDisabled);
```

File: src/include/platform/ConfigurationManager.h [hunk 22]
```diff
@@ -192,6 +197,9 @@ class ConfigurationManager
     ConfigurationManager(const ConfigurationManager &)  = delete;
     ConfigurationManager(const ConfigurationManager &&) = delete;
     ConfigurationManager & operator=(const ConfigurationManager &) = delete;
+
+    uint16_t gSecuredPort   = CHIP_PORT;
```
Comment (andy31415) [inline src/include/platform/ConfigurationManager.h:201]: configurationmanager seems to have no data for any  other settings and it looks like a CRTP that various platforms implement.

This is not the right place to add a temporary in-memory config.

File: src/app/server/Mdns.h (hunk 21 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: fixed

File: src/include/platform/ConfigurationManager.h (hunk 22 above)
Comment (chrisdecenzo) [reply src/include/platform/ConfigurationManager.h]: moved port config to Mdns and Server

File: src/app/server/Mdns.cpp [hunk 23]
```diff
@@ -78,6 +78,33 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
 
 } // namespace
 
+uint16_t gSecuredPort   = CHIP_PORT;
+uint16_t gUnsecuredPort = CHIP_UDC_PORT;
+
+/// Sets the secure Matter port
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:84]: Doccomments should go in the header. As they stand now, I would consider deleting them. It is obvious that 'SetSecuredPort' is  actually a "sets the secured port" (same for all the other comments below).

It would be even better to document what a secured port is and what an unsecured one is:  when is one used and when is another one used. Should this not be  named something like CommisionableNodePort() and the reverse?

 I am unclear what the reverse is ... the method signature is "bool commisionableNode" so it seems to be a flag. SetCommisionablePort and SetNotCommisionablePort sounds wrong :/

File: src/app/server/Mdns.h [hunk 24]
```diff
@@ -24,21 +24,38 @@ namespace chip {
 namespace app {
 namespace Mdns {
 
+enum class CommissioningMode
+{
+    kDisabled,                        // Commissioning Mode is disabled, CM=0, AC=0 in DNS-SD key/value pairs
+    kEnabled,                         // Basic Commissioning Mode, CM=1, AC=0 in DNS-SD key/value pairs
```
Comment (andy31415) [inline src/app/server/Mdns.h:30]: nit: consider kBasic and kEnhanced as constants to match the comments.

File: src/app/server/Server.cpp [hunk 25]
```diff
@@ -405,6 +415,9 @@ void InitServer(AppDelegate * delegate)
 
 // ESP32 and Mbed OS examples have a custom logic for enabling DNS-SD
 #if CHIP_DEVICE_CONFIG_ENABLE_MDNS && !CHIP_DEVICE_LAYER_TARGET_ESP32 && !CHIP_DEVICE_LAYER_TARGET_MBED
+    app::Mdns::SetSecuredPort(gSecuredServicePort);
```
Comment (andy31415) [inline src/app/server/Server.cpp:418]: why are the ports not set for esp32 and mbed? 

I understand the comment of saying thse platforms have custom start code (I believe they use platform dns) however ports seem to not depend on this.

**Review by andy31415 — APPROVED** [review]

**Review by hawk248 — APPROVED** [review]

File: src/app/server/Mdns.cpp (hunk 23 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: addressed via comment updates

File: src/app/server/Mdns.h (hunk 24 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.h]: done

File: src/app/server/Server.cpp (hunk 25 above)
Comment (chrisdecenzo) [reply src/app/server/Server.cpp:416]: fixed

**Comment by mspang** [comment]
@chrisdecenzo 

Smoke tests are failing after merging this PR. Can you help fix it ASAP ?

**Comment by franck-apple** [comment]
@andy31415 yes it is.

**Review by Zybi85 — APPROVED** [review]
